### PR TITLE
Android: CCFileUtils listFiles impl

### PR DESCRIPTION
**Pull request by PatriceJiang**
ref https://github.com/cocos2d/cocos2d-x/issues/18725

### CODE REVIEW

File: cocos/platform/android/CCFileUtils-android.cpp [hunk 1]
```diff
@@ -280,6 +280,45 @@ long FileUtilsAndroid::getFileSize(const std::string& filepath)
...
+    if(nullptr == dir) {
+        LOGD("... FileUtilsAndroid::failed to open dir %s", relativePath.c_str());
+        AAssetDir_close(dir);
+        return fileList;
+    }
+    const char *tmpDir = nullptr;
+    while((tmpDir = AAssetDir_getNextFileName(dir))!= nullptr)
+    {
+        string filepath(tmpDir);
+        if(isDirectoryExistInternal(filepath)) filepath += "/";
```
Comment (minggo) [inline cocos/platform/android/CCFileUtils-android.cpp:327]: Why using `isDirectoryExistInternal`? This function is to find a file, not a directory.

File: cocos/platform/android/CCFileUtils-android.cpp [hunk 2]
```diff
@@ -280,6 +280,45 @@ long FileUtilsAndroid::getFileSize(const std::string& filepath)
...
+    } else {
+        relativePath = fullPath;
+    }
+
+    if (nullptr == assetmanager) {
+        LOGD("... FileUtilsAndroid::assetmanager is nullptr");
+        return fileList;
+    }
+
+    auto *dir = AAssetManager_openDir(assetmanager, relativePath.c_str());
```
Comment (minggo) [inline cocos/platform/android/CCFileUtils-android.cpp:317]: Should consider OBB files?

File: cocos/platform/android/CCFileUtils-android.cpp (hunk 1 above)
Comment (PatriceJiang) [reply cocos/platform/android/CCFileUtils-android.cpp:327]: default implementation includes directories in result 
https://github.com/cocos2d/cocos2d-x/blob/3cad2a85d49fbd787c0c68e960fe0678b28bbc09/cocos/platform/CCFileUtils.cpp#L1453-L1459

File: cocos/base/ZipUtils.h [hunk 3]
```diff
@@ -265,6 +265,16 @@ typedef struct unz_file_info_s unz_file_info;
...
         bool fileExists(const std::string &fileName) const;
 
+
+        /**
+         * Get files and folders in pathname
+         *
+         * @param dirname
+         * @return
+         */
+        const std::vector<std::string> listFiles(const std::string &pathname) const;
```
Comment (minggo) [inline cocos/base/ZipUtils.h]: I think `const` is not needed if it doesn't return a reference.

File: cocos/base/ZipUtils.cpp [hunk 4]
```diff
@@ -609,6 +636,37 @@ bool ZipFile::fileExists(const std::string &fileName) const
     return ret;
 }
 
+const std::vector<std::string> ZipFile::listFiles(const std::string &pathname) const
+{
+
+    // filter files which `filename.startsWith(pathname)`
+    // then make each path unique
+
+    std::set<std::string> fileSet;
```
Comment (minggo) [inline cocos/base/ZipUtils.cpp:619]: Why use set? Does it have duplicate elements?

File: cocos/base/ZipUtils.cpp [hunk 5]
```diff
@@ -609,6 +636,37 @@ bool ZipFile::fileExists(const std::string &fileName) const
...
+        if(filename.substr(0, dirname.length()) == dirname)
+        {
+            std::string suffix = filename.substr(dirname.length());
+            const std::vector<std::string>& parts = splitFilePath(suffix);
+            if(parts.size() == 1)
+            {
+                fileSet.insert(suffix);
+            }else if(parts.size() > 1){
+                fileSet.insert(parts[0] + "/");
+            }
```
Comment (minggo) [inline cocos/base/ZipUtils.cpp]: `splitPath` is too complex here, we don't need to know every part content, just need to check if it has `/` or not?

File: cocos/base/ZipUtils.cpp (hunk 4 above)
Comment (PatriceJiang) [reply cocos/base/ZipUtils.cpp:619]: yes, `_data->fileList` caches all files in the zip file. 
```
a/b/c/
a/b/c/d
a/b/c/e
a/b/c/f/g
```
these paths will share same component `a/b/c/`

File: cocos/base/ZipUtils.cpp (hunk 5 above)
Comment (PatriceJiang) [reply cocos/base/ZipUtils.cpp]: ok.

**Review by minggo — APPROVED** [review]

**Comment by minggo** [comment]
@jianglong you can close and open the PR the trigger CI in future.

**Comment by PatriceJiang** [comment]
okay

**Pull request merged**